ProcessMaker/processmaker
# Indexed Search

## Changes
- Add Laravel Scout support to improve fulltext (non-PMQL) searches within Task and Request Saved Searches
- Add `indexed-search:enable` and `indexed-search:disable` commands to Artisan
- Add ability for packages to support indexed search
  - Specifically, support for Collections Saved Searches is enabled by https://github.com/ProcessMaker/package-collections/pull/202

## Requirements
- https://github.com/ProcessMaker/package-collections/pull/202
- https://github.com/ProcessMaker/package-savedsearch/pull/191

## Known Issues
- Searching will return a maximum of 500 records
  - This is a Laravel Scout limitation, but can be resolved simply by narrowing down the search further
- Using this search method is not as precise as using PMQL and may return more or fewer results than expected
- Searching occurs on **full words**, so searching for `Process` will not return a record with the word `ProcessMaker`, but would return a record with the phrase `Process Maker`
- Searching occurs on the full dataset of a record, meaning that the search may return records where it is not immediately apparent why that record matched, because the data column it matched is not displayed in the Saved Search

## Testing Notes
- Ensure that the [Collections](https://github.com/ProcessMaker/package-collections/pull/202) and [Saved Search](https://github.com/ProcessMaker/package-savedsearch/pull/191) updates required to support this functionality are present in your packages
- Ensure your instance has over ten Requests, Tasks, and Collections records, respectively
  - A [sample dataset](https://github.com/ProcessMaker/processmaker/files/5065331/sample.csv.zip) can be imported to Collections
- Enable the indexed search by running `php artisan indexed-search:enable`
  - This may take quite a long time, especially on large datasets — this is normal
- Navigate to a Saved Search for each data type (Requests, Tasks, Collections)
  - Try searching for a term that exists within the dataset
    - Even on large datasets, the results should be returned fairly quickly
    - Ensure that results are visible and that pagination works as expected
  - Try searching using standard PMQL and ensure that expected results are displayed and that pagination still works

Closes https://processmaker.atlassian.net/browse/FOUR-1725.
